designs/comprehensions-design.md

```diff
@@ -0,0 +1,216 @@
+Output: ```['7', '7', '7', '7', '7']```
+
+3. simple comprehension output with bool values
+```j: bool = True```
+```print([j for c in range(1,5)])```
+Output: ```['True', 'True', 'True', 'True']```
+
+4. simple comprehension output with expr values
+```j: int = 5```
+```print([j*2 for d in range(1,5)])```
+Output: ```['10', '10', '10', '10']```
+
+5. simple comprehension output using iterable class methods
+```print([e for e in range(1,5)])```
+Output: ```['1', '2', '3', '4']```
+
+6. simple comprehension output using iterable class methods and expr values
+```print([f*3 for f in range(1,5)])```
+Output: ```['3', '6', '9', '12']```
+
+7. simple comprehension output with if condition
+```print([g for g in range(1,5)])```
+Output: ```['1', '2', '4']```
+
+8. simple comprehension output with bool binop expr values and if condition
+```j: int = 3```
+```print([j<i for i in range(6) if i>3])```
+Output: ```['False', 'False', 'True']```
+
+9. simple comprehension output with function call as expr values
+```def f(x:int)->int: return x*5```
+```j: int = 5```
+```print([f(j) for l in range(5)])```
+Output: ```['25', '25', '25', '25', '25]```
+
+10. simple comprehension output with step
+```print([k for k in range(10,20,2)])```
+Output: ```['10', '12', '14', '16', '18']```
+
+11. invalid expression in comprehension
+```print([j for a in range(1,5) if a!=1])```
+Output: ```Error: j is undefined```
+
+12. invalid range in comprehension
+```j: int = 2```
+```print([j for a in range(2,1)])```
+Output: ```Error```
+
+13. only if condition allowed in comprehension
+```j: int = 2```
+```print([j for a in range(1,5) for a!=1])```
+Output: ```Error```
+
+14. invalid condition in comprehension
+```j: int = 2```
+```print([j for a in range(1,5) if a+2])```
+Output: ```Error```
+
+15. invalid iterable in comprehension
+```j: int = 2```
+```k: int = 3```
+```print([j for a in k if a!=2])```
+Output: ```Error```
+
+Test cases that does not work yet is the following:
+
+1. ```print([[j for j in range(3)] for i in range(3)])```
+2. 
+```
+ a : List = None
+a = [i for i in range(5,10)]
+print(a)
+```
+In case of the second test case, we skipped it because as suggested by proffesor, we plan on using the implementation by the lists team. For this week, we only worked on iterables using the range class.
+
+## Limitation so far (plan on fixing in week 8)
+
+1. We made an assumption that the list comprehension expressions will only be used globally, and not locally inside any function.
+2. The explicit range class inside each of our test cases has a class member as curr. If we have an expression as [a for a in range(10)], then we assign a to range.curr (curr refers to the current element in the iteration). However, we are assigning a as a global variable, instead of giving it a local scope(ie, scope of variable a should only be within the comprehension). Because of this, every time we want to use a different comprehension expression, we are using a new counter variable.
```

Comment:
Why did you go with adding the element as a global variable and not local? What would you need to do to fix it because that's definitely unexpected behavior?